### PR TITLE
[I] Errors - Rework errors management to be more precise

### PR DESCRIPTION
**Pull request by anthonyGuiguen**

### CODE REVIEW

File: Sources/ZetaPushNetwork/Client/ClientHelper.swift [hunk 1]
```diff
@@ -324,31 +294,35 @@ extension ClientHelper: CometdClientDelegate {
...
   
-  public func didDisconnected(error: Error?, from client: CometdClientContract) {
+  public func didDisconnected(error: CometDClientError?, from client: CometdClientContract) {
     log.zp.debug("ClientHelper Disconnected from Cometd server")
     connected = false
-    delegate?.onConnectionClosed(self)
+    if let error = error {
+      delegate?.onConnectionClosed(self, error: .connectionClosed(error: error))
+    } else {
+      delegate?.onConnectionClosed(self, error: nil)
```
Comment (StevenWatremez) [inline Sources/ZetaPushNetwork/Client/ClientHelper.swift:319]: Here you set nil to error, do you think about putting .unknown connection close error instead ?

File: Sources/ZetaPushNetwork/Client/ClientHelperDelegate.swift [hunk 2]
```diff
@@ -11,25 +11,27 @@ import Foundation
...
+  func onConnectionFailed(_ client: ClientHelper, error: ClientHelperError)
   func onConnectionEstablished(_ client: ClientHelper)
-  func onConnectionBroken(_ client: ClientHelper)
-  func onConnectionClosed(_ client: ClientHelper)
+  func onConnectionBroken(_ client: ClientHelper, error: ClientHelperError)
+  func onConnectionClosed(_ client: ClientHelper, error: ClientHelperError?)
   func onConnectionClosedAdviceReconnect(_ client: ClientHelper)
   func onSuccessfulHandshake(_ client: ClientHelper)
-  func onFailedHandshake(_ client: ClientHelper)
+  func onFailedHandshake(_ client: ClientHelper, error: ClientHelperError)
```
Comment (StevenWatremez) [inline Sources/ZetaPushNetwork/Client/ClientHelperDelegate.swift]: Here you put a ClientHelperError but for the API is it better compare to just Error ? I don't know the solution but it's  good to discussed together.

File: Sources/ZetaPushNetwork/Client/Server/ServerRemoteDataSourceError.swift [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    case .failed:
+      return ""
+    }
+  }
+  
+  func toNSError() -> NSError {
+    switch self {
+    case .canNotOpenURL(let url):
+      return NSError(domain: ErrorConstant.domain, code: 403, userInfo: [
+        NSLocalizedFailureReasonErrorKey: NSLocalizedString("Can not open URL", comment: ""),
```
Comment (StevenWatremez) [inline Sources/ZetaPushNetwork/Client/Server/ServerRemoteDataSourceError.swift]: here maybe it's a NSLocalizedDescriptionKey

File: Sources/ZetaPushNetwork/Client/Server/ServerRemoteDataSourceError.swift [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    case .canNotOpenURL:
+      return "ERROR_CLIENT_HELPER_CAN_NOT_OPEN_URL"
+    case .response:
+      return "ERROR_CLIENT_HELPER_RESPONSE"
+    case .unknown:
+      return "ERROR_CLIENT_HELPER_UNKNOWN"
+    case .serversNotFound:
+      return "ERROR_CLIENT_HELPER_SERVERS_NOT_FOUND"
+    case .failed:
+      return ""
```
Comment (StevenWatremez) [inline Sources/ZetaPushNetwork/Client/Server/ServerRemoteDataSourceError.swift]: why here you don't create a code ?

File: Sources/ZetaPushNetwork/Client/ClientHelper.swift (hunk 1 above)
Comment (anthonyGuiguen) [reply Sources/ZetaPushNetwork/Client/ClientHelper.swift:319]: A connection can be closed without error

File: Sources/ZetaPushNetwork/Client/Server/ServerRemoteDataSourceError.swift (hunk 4 above)
Comment (anthonyGuiguen) [reply Sources/ZetaPushNetwork/Client/Server/ServerRemoteDataSourceError.swift]: Yes I forgot

**Review by StevenWatremez — APPROVED** [review]

**Review by jgoas — APPROVED** [review]

**Pull request merged**